src/Player/PlayerState/PlayerStateHack.cpp

```diff
@@ -0,0 +1,276 @@
+    }
+    s32 tmp = mHackDemoStartLength - 4;
+    s32 nerveStep = tmp < 0 ? 0 : tmp;
+    f32 nerveEastRate;
+    if (!al::isFirstStep(this)) {
+        if (!al::isLessStep(this, nerveStep)) {
+            nerveEastRate = al::calcNerveEaseInValue(this, nerveStep, nerveStep + 5, 1.0, 0.02);
+        } else {
+            nerveEastRate = al::calcNerveEaseOutValue(this, 0, 5, 0.02, 1.0);
+        }
+
+    } else {
+        nerveEastRate = 1.0f;
+    }
+    if (al::isStep(this, mHackDemoStartLength + 9)) {
+        mHackCap->noticeHackMarioEnter();
+    }
+    tmp = mHackDemoStartLength - 15;
+    nerveStep = tmp < 0 ? 0 : tmp;
+    if (al::isStep(this, nerveStep)) {
+        mHackCap->emitHackStartEffect();
+    }
+    if (al::isLessStep(this, mHackDemoStartLength)) {
+        f32 nerveRate = al::calcNerveRate(this, mHackDemoStartLength);
+        sead::Vector3f* capTrans = &al::getTrans(mHackCap);
+        sead::Vector3f lerpedTrans = {0.0f, 0.0f, 0.0f};
+        al::lerpVec(&lerpedTrans, mTrans, al::getTrans(hackActor), nerveRate);
+        al::resetPosition(getActor(), lerpedTrans);
+        sead::Vector3f beforeLerpSensorTrans = mNewSensorTrans;
+        al::lerpVec(&mNewSensorTrans, mSensorTrans, *capTrans, nerveRate);
+
+        sead::Vector3CalcCommon<f32>::multScalarAdd(mNewSensorTrans, sinf(nerveRate * 0.01745329f * 180.0f) * -400.0f,
+                                                    al::getGravity(getActor()), mNewSensorTrans);
+        sead::Vector3CalcCommon<f32>::sub(frontDir, mNewSensorTrans, beforeLerpSensorTrans);
+
+        if (!al::tryNormalizeOrZero(&frontDir)) {
+            sead::Vector3f* actorTrans = &al::getTrans(actor);
+            sead::Vector3CalcCommon<f32>::sub(frontDir, *capTrans, *actorTrans);
+            if (!al::tryNormalizeOrZero(&frontDir)) {
+                al::calcFrontDir(&frontDir, actor);
+            }
+        }
+        upDir = -al::getGravity(actor);
+        if (al::isParallelDirection(upDir, frontDir, 0.01)) {
+            al::calcFrontDir(&upDir, actor);
+            if (al::isParallelDirection(upDir, frontDir, 0.01)) {
+                al::calcUpDir(&upDir, actor);
+            }
+        }
+        al::makeMtxFrontUpPos(&mPossessTraceMtx, frontDir, upDir, mNewSensorTrans);
+        reverseFront = -frontDir;
+        downDir = -upDir;
+        cameraFront = {0.0f, 0.0f, 0.0f};
+
+        al::calcCameraFront(&cameraFront, actor, 0);
+        if (al::isParallelDirection(downDir, cameraFront, 0.01)) {
+            if (al::isReverseDirection(downDir, cameraFront, 0.01)) {
+                downDir *= -1;
+            }
+        } else if (!al::isParallelDirection(cameraFront, reverseFront, 0.01)) {
+            sead::Quatf quat = sead::Quatf::unit;
+            reverseCameraFront = -cameraFront;
+
+            al::makeQuatAxisRotation(&quat, downDir, reverseCameraFront, reverseFront, 1.0);
+            sead::Vector3CalcCommon<f32>::rotate(downDir, quat, downDir);
```

Comment:
This does not exist (yet)?

Reply:
correct, it should, but i haven't been able to get a matching impl yet. i do have *an* impl in seadVectorCalcCommon, but its not perfect yet.

Reply:
Well, uhh... don't try to PR non-compiling code, please 😅

Reply:
it does compile, its just not matching yet.